### PR TITLE
Add log and factorial function

### PR DESCRIPTION
**Pull request by esbenbjerre**
Add log and factorial to Int32 as per #1246.

### CODE REVIEW

**Comment by magnus-madsen** [comment]
Looks reasonable. Please add 3-5 test cases for each function in `TestInt32.flix`.

**Comment by magnus-madsen** [comment]
Question: Do you want a separate `log2` function for the complexity bounds? And can this specific one be implemented more efficiently than the general log?

**Comment by esbenbjerre** [comment]
> Looks reasonable. Please add 3-5 test cases for each function in `TestInt32.flix`.

Will do.



> Question: Do you want a separate `log2` function for the complexity bounds? And can this specific one be implemented more efficiently than the general log?

I'm not sure, but I will look into it :)

**Comment by esbenbjerre** [comment]
The integer binary logarithm can be efficiently calculated using a few bit tricks.

However, it seems my arbitrary-base integer logarithm is flawed.
For example, `Int32.log(1000, 10)` returns `2` but should return `3`.
I suspect this is due to some floating-point error.
There seems to be no pretty fix here, so I have opted to remove it for now, as I only need log2 for the complexities.

File: main/src/library/Int32.flix [hunk 1]
```diff
@@ -364,6 +374,11 @@ namespace Int32 {
         let maxi32 = Int16.toInt32(maximum);
         valueOf(clamp(x, mini32, maxi32)).shortValue() as & Pure
 
+    ///
```
Comment (magnus-madsen) [inline main/src/library/Int32.flix:377]: Helper function should be right after the function itself.

(Once we have letrec, we will want to inline such helper functions, and having them close will be a useful aid).

File: main/src/library/Int32.flix [hunk 2]
```diff
@@ -235,6 +235,16 @@ namespace Int32 {
     ///
     pub def flipBit(x: Int32, position: Int32): Int32 = x ^^^ (1 <<< position)
 
+    ///
+    /// Returns the integer binary logarithm of `x`.
+    ///
+    pub def log2(x: Int32): Int32 = size() - numberOfLeadingZeros(x) - 1
```
Comment (magnus-madsen) [inline main/src/library/Int32.flix:241]: Cool!

File: main/test/ca/uwaterloo/flix/library/TestInt32.flix [hunk 3]
```diff
@@ -594,7 +594,39 @@ namespace TestInt32 {
...
+    /////////////////////////////////////////////////////////////////////////////
+
+    @test
+    def log201(): Bool = Int32.log2(10) == 3
+
+    @test
+    def log202(): Bool = Int32.log2(10000) == 13
+
+    @test
+    def log203(): Bool = Int32.log2(8) == 3
```
Comment (magnus-madsen) [inline main/test/ca/uwaterloo/flix/library/TestInt32.flix:609]: Can I ask you to add just a few more tests for both?

File: main/src/library/Int32.flix (hunk 2 above)
Comment (esbenbjerre) [reply main/src/library/Int32.flix]: I've simplified it further after seeing that `numberOfLeadingZeros` is implemented using `highestOneBitPosition`.

File: main/test/ca/uwaterloo/flix/library/TestInt32.flix (hunk 3 above)
Comment (esbenbjerre) [reply main/test/ca/uwaterloo/flix/library/TestInt32.flix:609]: Yes!

**Comment by magnus-madsen** [comment]
Merged! Thanks

**Pull request merged**